### ISSUE TITLE
Sending a client certificate causes an error, and no response

### ISSUE THREAD

**Issue by makeworld-the-better-one** [michael-lazar/jetforce#20]
I tried to send a client cert to my server and the page wouldn't load. This is the error left in the log:
```
May 18 12:02:56 jupiter jetforce[21138]: SSL handshake failed on verifying the certificate
May 18 12:02:56 jupiter jetforce[21138]: protocol: <asyncio.sslproto.SSLProtocol object at 0x7f01f90c9400>
May 18 12:02:56 jupiter jetforce[21138]: transport: <_SelectorSocketTransport fd=9 read=polling write=<idle, bufsize=0>>
May 18 12:02:56 jupiter jetforce[21138]: Traceback (most recent call last):
May 18 12:02:56 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 625, in _on_handshake_complete
May 18 12:02:56 jupiter jetforce[21138]:     raise handshake_exc
May 18 12:02:56 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
May 18 12:02:56 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:02:56 jupiter jetforce[21138]:   File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
May 18 12:02:56 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:02:56 jupiter jetforce[21138]: ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
May 18 12:02:56 jupiter jetforce[21138]: SSL error in data received
May 18 12:02:56 jupiter jetforce[21138]: protocol: <asyncio.sslproto.SSLProtocol object at 0x7f01f90c9400>
May 18 12:02:56 jupiter jetforce[21138]: transport: <_SelectorSocketTransport closing fd=9 read=idle write=<idle, bufsize=0>>
May 18 12:02:56 jupiter jetforce[21138]: Traceback (most recent call last):
May 18 12:02:56 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 526, in data_received
May 18 12:02:56 jupiter jetforce[21138]:     ssldata, appdata = self._sslpipe.feed_ssldata(data)
May 18 12:02:56 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
May 18 12:02:56 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:02:56 jupiter jetforce[21138]:   File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
May 18 12:02:56 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:02:56 jupiter jetforce[21138]: ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
May 18 12:03:00 jupiter jetforce[21138]: SSL handshake failed on verifying the certificate
May 18 12:03:00 jupiter jetforce[21138]: protocol: <asyncio.sslproto.SSLProtocol object at 0x7f01f90a3a58>
May 18 12:03:00 jupiter jetforce[21138]: transport: <_SelectorSocketTransport fd=9 read=polling write=<idle, bufsize=0>>
May 18 12:03:00 jupiter jetforce[21138]: Traceback (most recent call last):
May 18 12:03:00 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 625, in _on_handshake_complete
May 18 12:03:00 jupiter jetforce[21138]:     raise handshake_exc
May 18 12:03:00 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
May 18 12:03:00 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:03:00 jupiter jetforce[21138]:   File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
May 18 12:03:00 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:03:00 jupiter jetforce[21138]: ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
May 18 12:03:00 jupiter jetforce[21138]: SSL error in data received
May 18 12:03:00 jupiter jetforce[21138]: protocol: <asyncio.sslproto.SSLProtocol object at 0x7f01f90a3a58>
May 18 12:03:00 jupiter jetforce[21138]: transport: <_SelectorSocketTransport closing fd=9 read=idle write=<idle, bufsize=0>>
May 18 12:03:00 jupiter jetforce[21138]: Traceback (most recent call last):
May 18 12:03:00 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 526, in data_received
May 18 12:03:00 jupiter jetforce[21138]:     ssldata, appdata = self._sslpipe.feed_ssldata(data)
May 18 12:03:00 jupiter jetforce[21138]:   File "/usr/lib/python3.7/asyncio/sslproto.py", line 189, in feed_ssldata
May 18 12:03:00 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:03:00 jupiter jetforce[21138]:   File "/usr/lib/python3.7/ssl.py", line 763, in do_handshake
May 18 12:03:00 jupiter jetforce[21138]:     self._sslobj.do_handshake()
May 18 12:03:00 jupiter jetforce[21138]: ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
```

The behaviour I would expect is that the client cert is just ignored, or not verified, etc. This is problem for me right now because the Bombadillo client sends your specified client cert with every request, so when I add a client cert, maybe for astrobotany or something, all Jetforce servers break.

**Comment by michael-lazar** [comment]
So because of limitations in the python OpenSSL bindings, there are only two behaviors to choose from:

- The server doesn't request the client certificate under any circumstances.
- The server requests the client certificate. If a certificate is presented, it must be validated.

If you use specify either the ``--tls-cafile`` or ``--tls-capath`` options when you launch the server, it will trigger the second behavior. I appreciate that this is a problem with self-signed certificates and I'm working on using the alternate PyOpenSSL bindings to work around the issue. For now, if you want your server to work with random client certificates you can disable certificate validation all together by removing the ``--tls-cafile`` argument. I think that most servers should not be using this option anyway, because I don't know of any gemini servers that use client CA's outside of astrobotany.

**Comment by makeworld-the-better-one** [comment]
> For now, if you want your server to work with random client certificates you can disable certificate validation all together by removing the `--tls-cafile` argument. 

My server doesn't have this specified, that was why I filed the issue. I expected the server to just ignore the client cert, but instead it threw an error.

Here are my server options, I think they're pretty mundane.
```
ExecStart=/usr/local/bin/jetforce \
    --host 0.0.0.0 \
    --hostname <redacted> \
    --dir /var/gemini \
    --tls-certfile /etc/ssl/caddy/acme/acme-v02.api.letsencrypt.org/sites/<redacted>/<redacted>.crt \
    --tls-keyfile /etc/ssl/caddy/acme/acme-v02.api.letsencrypt.org/sites/<redacted>/<redacted>.key
```

**Comment by michael-lazar** [comment]
Hmm you're right, here's the code (I must have re-written it back when I was trying to get client certificates working).

```python
    context = ssl.SSLContext()
    context.verify_mode = ssl.CERT_OPTIONAL
    context.load_cert_chain(certfile, keyfile)

    if not cafile and not capath:
        # Load from the system's default client CA directory
        context.load_default_certs(purpose=ssl.Purpose.CLIENT_AUTH)
    else:
        # Use a custom CA for validating client certificates
        context.load_verify_locations(cafile, capath)
```

This would probably make more sense as

```python
    context = ssl.SSLContext()
    context.load_cert_chain(certfile, keyfile)

    if not cafile and not capath:
        context.verify_mode = ssl.CERT_NONE
    else:
        # Use a custom CA for validating client certificates
        context.verify_mode = ssl.CERT_OPTIONAL
        context.load_verify_locations(cafile, capath)
```

**Comment by makeworld-the-better-one** [comment]
I don't know all the ins and outs about this, but that looks right! Thanks. Do you think this can be put in the codebase for the next release?

**Comment by michael-lazar** [comment]
The next release will handle this better. Client certificates will still be requested by OpenSSL, but even if the certificate validation fails, the server will still allow the request to continue. The validation result from OpenSSL will be passed into the ``TLS_CLIENT_VERIFIED`` GCI variable so that individual CGI scripts can decide if they want to accept unverified certificates or not.

**Comment by makeworld-the-better-one** [comment]
Hi, did this fix make it into v2.3.1?

**Comment by michael-lazar** [comment]
- v0.2.3 was a security patch only and did not include a fix for this
- v0.3.0 is currently in beta and should fix this behavior

You can try it out now if you want to, I would actually appreciate having some extra eyes on it.

```
pip install jetforce==0.3.0b5
```
